terraform crashing on validate

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

If your issue relates to a specific Terraform provider, please open it in the provider's own repository. The index of providers is at https://github.com/terraform-providers .
-->

### Terraform Version
<!---
Run `terraform -v` to show the version, and paste the result between the ``` marks below.

If you are not running the latest version of Terraform, please try upgrading because your issue may have already been fixed.
-->

```
terraform --version
Terraform v0.12.2
+ provider.google v2.8.0

...
```

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For large Terraform configs, please use a service like Dropbox and share a link to the ZIP file. For security, you can also encrypt the files using our GPG public key.
-->
tf file below
```hcl
variable "zone" {
    type = string
}
variable "names" {
    type = list(string)
}
variable "records" {
    type = "map"
}
variable "ttl" {
    type = number
}

resource "google_dns_managed_zone" "prod" {
    name = var.zone
    dns_name = var.zone
}

resource "google_dns_record_set" "prod" {
    dynamic "name" {
        for_each = var.names
        content {
            dynamic "record" {
                for_each = var.records[name.value]
                content {
                    name = name.value
                    type = record.key
                    ttl = var.ttl
                    managed_zone = google_dns_managed_zone.prod.name
                    rrdatas = record.value[*]
                }
            }
        }
    }
}

...
```
tfvars file below
```hcl
ttl = 1800
records = {
    mydomain.com. = {
        "MX" = [(0,"mxmail.mydomain.com.")]
        "A" = [127.0.0.1]
        "NS" = [ns1.mydomain.com,ns2.mydomain.com]
    }
    *.mydomain.com. = {
        "MX" = [(0,"mxmail.mydomain.com.")]
        "A" = [127.0.0.1]
    }
}
names = [mydomain.com, *.mydomain.com]


...
```

### Debug Output
<!--
Full debug output can be obtained by running Terraform with the environment variable `TF_LOG=trace`. Please create a GitHub Gist containing the debug output. Please do _not_ paste the debug output in the issue, since debug output is long.

Debug output may contain sensitive information. Please review it before posting publicly, and if you are concerned feel free to encrypt the files using the HashiCorp security public key.
-->

```
2019/06/14 11:45:10 [INFO] Terraform version: 0.12.2  
2019/06/14 11:45:10 [INFO] Go runtime version: go1.12.4
2019/06/14 11:45:10 [INFO] CLI args: []string{"/usr/bin/terraform", "validate", "-var-file=test.tfvars"}
2019/06/14 11:45:10 [DEBUG] Attempting to open CLI config file: /home/jparker/.terraformrc
2019/06/14 11:45:10 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2019/06/14 11:45:10 [INFO] CLI command args: []string{"validate", "-var-file=test.tfvars"}
2019/06/14 11:45:10 [DEBUG] checking for provider in "."
2019/06/14 11:45:10 [DEBUG] checking for provider in "/usr/bin"
2019/06/14 11:45:10 [DEBUG] checking for provider in ".terraform/plugins/linux_amd64"
2019/06/14 11:45:10 [DEBUG] found provider "terraform-provider-google_v2.8.0_x4"
2019/06/14 11:45:10 [DEBUG] found valid plugin: "google", "2.8.0", "/home/jparker/repos/terraform/build/zones/.terraform/plugins/linux_amd64/terraform-provider-google_v2.8.0_x4"
2019/06/14 11:45:10 [DEBUG] checking for provisioner in "."
2019/06/14 11:45:10 [DEBUG] checking for provisioner in "/usr/bin"
2019/06/14 11:45:10 [DEBUG] checking for provisioner in ".terraform/plugins/linux_amd64"
2019/06/14 11:45:10 [TRACE] terraform.NewContext: starting
2019/06/14 11:45:10 [TRACE] terraform.NewContext: resolving provider version selections
2019/06/14 11:45:11 [TRACE] terraform.NewContext: loading provider schemas
2019/06/14 11:45:11 [TRACE] LoadSchemas: retrieving schema for provider type "google"
2019-06-14T11:45:11.120-0500 [INFO]  plugin: configuring client automatic mTLS
2019-06-14T11:45:11.155-0500 [DEBUG] plugin: starting plugin: path=/home/jparker/repos/terraform/build/zones/.terraform/plugins/linux_amd64/terraform-provider-google_v2.8.0_x4 args=[/home/jparker/repos/terraform/build/zones/.terraform/plugins/linux_amd64/terraform-provider-google_v2.8.0_x4]
2019-06-14T11:45:11.155-0500 [DEBUG] plugin: plugin started: path=/home/jparker/repos/terraform/build/zones/.terraform/plugins/linux_amd64/terraform-provider-google_v2.8.0_x4 pid=13910
2019-06-14T11:45:11.155-0500 [DEBUG] plugin: waiting for RPC address: path=/home/jparker/repos/terraform/build/zones/.terraform/plugins/linux_amd64/terraform-provider-google_v2.8.0_x4
2019-06-14T11:45:11.164-0500 [INFO]  plugin.terraform-provider-google_v2.8.0_x4: configuring server automatic mTLS: timestamp=2019-06-14T11:45:11.164-0500
2019-06-14T11:45:11.200-0500 [DEBUG] plugin.terraform-provider-google_v2.8.0_x4: plugin address: address=/tmp/plugin383768069 network=unix timestamp=2019-06-14T11:45:11.200-0500
2019-06-14T11:45:11.200-0500 [DEBUG] plugin: using plugin: version=5
2019/06/14 11:45:11 [TRACE] GRPCProvider: GetSchema
2019/06/14 11:45:11 [TRACE] GRPCProvider: Close
2019-06-14T11:45:11.297-0500 [DEBUG] plugin: plugin process exited: path=/home/jparker/repos/terraform/build/zones/.terraform/plugins/linux_amd64/terraform-provider-google_v2.8.0_x4 pid=13910
2019-06-14T11:45:11.297-0500 [DEBUG] plugin: plugin exited
2019/06/14 11:45:11 [TRACE] terraform.NewContext: complete
2019/06/14 11:45:11 [INFO] terraform: building graph: GraphTypeValidate
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.ConfigTransformer
2019/06/14 11:45:11 [TRACE] ConfigTransformer: Starting for path: 
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.ConfigTransformer with new graph:
google_dns_managed_zone.prod - *terraform.NodeValidatableResource
google_dns_record_set.prod - *terraform.NodeValidatableResource
------
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.LocalTransformer
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.LocalTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.OutputTransformer
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.OutputTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.OrphanResourceInstanceTransformer
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.OrphanResourceInstanceTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.StateTransformer
2019/06/14 11:45:11 [TRACE] StateTransformer: state is empty, so nothing to do
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.StateTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.OrphanOutputTransformer
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.OrphanOutputTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.AttachResourceConfigTransformer
2019/06/14 11:45:11 [TRACE] AttachResourceConfigTransformer: attaching to "google_dns_managed_zone.prod" (*terraform.NodeValidatableResource) config from zones.tf:14,1-42
2019/06/14 11:45:11 [TRACE] AttachResourceConfigTransformer: attaching to "google_dns_record_set.prod" (*terraform.NodeValidatableResource) config from zones.tf:19,1-40
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.AttachResourceConfigTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.AttachStateTransformer
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.AttachStateTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.RootVariableTransformer
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.RootVariableTransformer with new graph:
google_dns_managed_zone.prod - *terraform.NodeValidatableResource
google_dns_record_set.prod - *terraform.NodeValidatableResource
var.names - *terraform.NodeRootVariable
var.records - *terraform.NodeRootVariable
var.ttl - *terraform.NodeRootVariable
var.zone - *terraform.NodeRootVariable
------
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.MissingProvisionerTransformer
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.MissingProvisionerTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.ProvisionerTransformer
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.ProvisionerTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.ModuleVariableTransformer
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.ModuleVariableTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.graphTransformerMulti
2019/06/14 11:45:11 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.ProviderConfigTransformer
2019/06/14 11:45:11 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.ProviderConfigTransformer with new graph:
google_dns_managed_zone.prod - *terraform.NodeValidatableResource
google_dns_record_set.prod - *terraform.NodeValidatableResource
var.names - *terraform.NodeRootVariable
var.records - *terraform.NodeRootVariable
var.ttl - *terraform.NodeRootVariable
var.zone - *terraform.NodeRootVariable
------
2019/06/14 11:45:11 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.MissingProviderTransformer
2019/06/14 11:45:11 [DEBUG] adding implicit provider configuration provider.google, implied first by google_dns_managed_zone.prod
2019/06/14 11:45:11 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.MissingProviderTransformer with new graph:
google_dns_managed_zone.prod - *terraform.NodeValidatableResource
google_dns_record_set.prod - *terraform.NodeValidatableResource
provider.google - *terraform.NodeApplyableProvider
var.names - *terraform.NodeRootVariable
var.records - *terraform.NodeRootVariable
var.ttl - *terraform.NodeRootVariable
var.zone - *terraform.NodeRootVariable
------
2019/06/14 11:45:11 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.ProviderTransformer
2019/06/14 11:45:11 [TRACE] ProviderTransformer: google_dns_managed_zone.prod is provided by provider.google or inherited equivalent
2019/06/14 11:45:11 [TRACE] ProviderTransformer: google_dns_record_set.prod is provided by provider.google or inherited equivalent
2019/06/14 11:45:11 [TRACE] ProviderTransformer: exact match for provider.google serving google_dns_record_set.prod
2019/06/14 11:45:11 [DEBUG] ProviderTransformer: "google_dns_record_set.prod" (*terraform.NodeValidatableResource) needs provider.google
2019/06/14 11:45:11 [TRACE] ProviderTransformer: exact match for provider.google serving google_dns_managed_zone.prod
2019/06/14 11:45:11 [DEBUG] ProviderTransformer: "google_dns_managed_zone.prod" (*terraform.NodeValidatableResource) needs provider.google
2019/06/14 11:45:11 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.ProviderTransformer with new graph:
google_dns_managed_zone.prod - *terraform.NodeValidatableResource
  provider.google - *terraform.NodeApplyableProvider
google_dns_record_set.prod - *terraform.NodeValidatableResource
  provider.google - *terraform.NodeApplyableProvider
provider.google - *terraform.NodeApplyableProvider
var.names - *terraform.NodeRootVariable
var.records - *terraform.NodeRootVariable
var.ttl - *terraform.NodeRootVariable
var.zone - *terraform.NodeRootVariable
------
2019/06/14 11:45:11 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.PruneProviderTransformer
2019/06/14 11:45:11 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.PruneProviderTransformer (no changes)
2019/06/14 11:45:11 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.ParentProviderTransformer
2019/06/14 11:45:11 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.ParentProviderTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.graphTransformerMulti with new graph:
google_dns_managed_zone.prod - *terraform.NodeValidatableResource
  provider.google - *terraform.NodeApplyableProvider
google_dns_record_set.prod - *terraform.NodeValidatableResource
  provider.google - *terraform.NodeApplyableProvider
provider.google - *terraform.NodeApplyableProvider
var.names - *terraform.NodeRootVariable
var.records - *terraform.NodeRootVariable
var.ttl - *terraform.NodeRootVariable
var.zone - *terraform.NodeRootVariable
------
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.RemovedModuleTransformer
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.RemovedModuleTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.AttachSchemaTransformer
2019/06/14 11:45:11 [TRACE] AttachSchemaTransformer: attaching provider config schema to provider.google
2019/06/14 11:45:11 [TRACE] AttachSchemaTransformer: attaching resource schema to google_dns_managed_zone.prod
2019/06/14 11:45:11 [TRACE] AttachSchemaTransformer: attaching resource schema to google_dns_record_set.prod
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.AttachSchemaTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.ReferenceTransformer
panic: not a collection type

goroutine 1 [running]:
github.com/zclconf/go-cty/cty.Type.ElementType(0x21eabe0, 0xc000040469, 0xc000041020, 0x4)
        /opt/gopath/pkg/mod/github.com/zclconf/go-cty@v0.0.0-20190516203816-4fecf87372ec/cty/collection.go:25 +0xd2
github.com/hashicorp/terraform/lang/blocktoattr.walkVariables(0x21e9860, 0xc0002fd550, 0x0, 0xc00033f201, 0x21e9860, 0xc0002fd550, 0xc000319de0, 0x40b849, 0x73, 0xc00015afac)
        /opt/gopath/src/github.com/hashicorp/terraform/lang/blocktoattr/variables.go:37 +0x45e
github.com/hashicorp/terraform/lang/blocktoattr.ExpandedVariables(...)
        /opt/gopath/src/github.com/hashicorp/terraform/lang/blocktoattr/variables.go:24
github.com/hashicorp/terraform/lang.ReferencesInBlock(0x21e9860, 0xc0002fd550, 0xc000319de0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x7f3ffd1cd008)
        /opt/gopath/src/github.com/hashicorp/terraform/lang/references.go:68 +0x66
github.com/hashicorp/terraform/terraform.(*NodeAbstractResource).References(0xc000483ce0, 0x0, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/node_resource_abstract.go:190 +0x251
github.com/hashicorp/terraform/terraform.NewReferenceMap(0xc000113b20, 0x7, 0x7, 0x7)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/transform_reference.go:402 +0x24a
github.com/hashicorp/terraform/terraform.(*ReferenceTransformer).Transform(0x32e0c40, 0xc0000bf630, 0x14, 0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/transform_reference.go:73 +0x71
github.com/hashicorp/terraform/terraform.(*BasicGraphBuilder).Build(0xc00033f798, 0x0, 0x0, 0x0, 0x32c0f80, 0xc00033f7c8, 0x4e5650, 0xc0000be000)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/graph_builder.go:50 +0x477
github.com/hashicorp/terraform/terraform.(*PlanGraphBuilder).Build(0xc0001127e0, 0x0, 0x0, 0x0, 0x1, 0x0, 0x1a6f440, 0xc0004078c0)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/graph_builder_plan.go:65 +0xbc
github.com/hashicorp/terraform/terraform.(*Context).Graph(0xc0000180f0, 0x6, 0xc00033f81e, 0xc00033f900, 0x1, 0x1, 0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/context.go:292 +0x1b8
github.com/hashicorp/terraform/terraform.(*Context).Validate(0xc0000180f0, 0x0, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/context.go:670 +0xf3
github.com/hashicorp/terraform/command.(*ValidateCommand).validate(0xc0003ef500, 0xc0003fc150, 0x29, 0x0, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/command/validate.go:114 +0x4f9
github.com/hashicorp/terraform/command.(*ValidateCommand).Run(0xc0003ef500, 0xc00003a1a0, 0x1, 0x1, 0xc000091b60)
        /opt/gopath/src/github.com/hashicorp/terraform/command/validate.go:70 +0x6c2
github.com/mitchellh/cli.(*CLI).Run(0xc0000fe640, 0xc0000fe640, 0xc000531d90, 0x1)
        /opt/gopath/pkg/mod/github.com/mitchellh/cli@v1.0.0/cli.go:255 +0x1f1
main.wrappedMain(0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/main.go:223 +0xaff
main.realMain(0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/main.go:100 +0xb4
main.main()
        /opt/gopath/src/github.com/hashicorp/terraform/main.go:36 +0x3b
panic: not a collection type

goroutine 1 [running]:
github.com/zclconf/go-cty/cty.Type.ElementType(0x21eabe0, 0xc000040469, 0xc000041020, 0x4)
        /opt/gopath/pkg/mod/github.com/zclconf/go-cty@v0.0.0-20190516203816-4fecf87372ec/cty/collection.go:25 +0xd2
github.com/hashicorp/terraform/lang/blocktoattr.walkVariables(0x21e9860, 0xc0002fd550, 0x0, 0xc00033f201, 0x21e9860, 0xc0002fd550, 0xc000319de0, 0x40b849, 0x73, 0xc00015afac)
        /opt/gopath/src/github.com/hashicorp/terraform/lang/blocktoattr/variables.go:37 +0x45e
github.com/hashicorp/terraform/lang/blocktoattr.ExpandedVariables(...)
        /opt/gopath/src/github.com/hashicorp/terraform/lang/blocktoattr/variables.go:24
github.com/hashicorp/terraform/lang.ReferencesInBlock(0x21e9860, 0xc0002fd550, 0xc000319de0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x7f3ffd1cd008)
        /opt/gopath/src/github.com/hashicorp/terraform/lang/references.go:68 +0x66
github.com/hashicorp/terraform/terraform.(*NodeAbstractResource).References(0xc000483ce0, 0x0, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/node_resource_abstract.go:190 +0x251
github.com/hashicorp/terraform/terraform.NewReferenceMap(0xc000113b20, 0x7, 0x7, 0x7)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/transform_reference.go:402 +0x24a
github.com/hashicorp/terraform/terraform.(*ReferenceTransformer).Transform(0x32e0c40, 0xc0000bf630, 0x14, 0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/transform_reference.go:73 +0x71
github.com/hashicorp/terraform/terraform.(*BasicGraphBuilder).Build(0xc00033f798, 0x0, 0x0, 0x0, 0x32c0f80, 0xc00033f7c8, 0x4e5650, 0xc0000be000)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/graph_builder.go:50 +0x477
github.com/hashicorp/terraform/terraform.(*PlanGraphBuilder).Build(0xc0001127e0, 0x0, 0x0, 0x0, 0x1, 0x0, 0x1a6f440, 0xc0004078c0)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/graph_builder_plan.go:65 +0xbc
github.com/hashicorp/terraform/terraform.(*Context).Graph(0xc0000180f0, 0x6, 0xc00033f81e, 0xc00033f900, 0x1, 0x1, 0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/context.go:292 +0x1b8
github.com/hashicorp/terraform/terraform.(*Context).Validate(0xc0000180f0, 0x0, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/context.go:670 +0xf3
github.com/hashicorp/terraform/command.(*ValidateCommand).validate(0xc0003ef500, 0xc0003fc150, 0x29, 0x0, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/command/validate.go:114 +0x4f9
github.com/hashicorp/terraform/command.(*ValidateCommand).Run(0xc0003ef500, 0xc00003a1a0, 0x1, 0x1, 0xc000091b60)
        /opt/gopath/src/github.com/hashicorp/terraform/command/validate.go:70 +0x6c2
github.com/mitchellh/cli.(*CLI).Run(0xc0000fe640, 0xc0000fe640, 0xc000531d90, 0x1)
        /opt/gopath/pkg/mod/github.com/mitchellh/cli@v1.0.0/cli.go:255 +0x1f1
main.wrappedMain(0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/main.go:223 +0xaff
main.realMain(0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/main.go:100 +0xb4
main.main()
        /opt/gopath/src/github.com/hashicorp/terraform/main.go:36 +0x3b
```

### Crash Output
<!--
If the console output indicates that Terraform crashed, please share a link to a GitHub Gist containing the output of the `crash.log` file.
-->

### Expected Behavior
<!--
What should have happened?
-->

### Actual Behavior
<!--
What actually happened?
-->

### Steps to Reproduce
<!--
Please list the full steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform apply`
-->
Use provided files
terraform init
terraform validate

### Additional Context
<!--
Are there anything atypical about your situation that we should know? For example: is Terraform running in a wrapper script or in a CI system? Are you passing any unusual command line options or environment variables to opt-in to non-default behavior?
-->

### References
<!--
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:

- #6017

-->

--- a/lang/blocktoattr/variables.go
+++ b/lang/blocktoattr/variables.go
@@ -34,6 +34,8 @@ func walkVariables(node dynblock.WalkVariablesNode, body hcl.Body, schema *confi
 		if blockS, exists := schema.BlockTypes[child.BlockTypeName]; exists {
 			vars = append(vars, walkVariables(child.Node, child.Body(), &blockS.Block)...)
 		} else if attrS, exists := schema.Attributes[child.BlockTypeName]; exists && attrS.Type.IsCollectionType() && attrS.Type.ElementType().IsObjectType() {
+			// ☝️Check for collection type before element type, because if this is a mis-placed reference,
+			// a panic here will prevent other useful diags from being elevated to show the user what to fix
 			synthSchema := SchemaForCtyElementType(attrS.Type.ElementType())
 			vars = append(vars, walkVariables(child.Node, child.Body(), synthSchema)...)
 		}

--- a/lang/blocktoattr/variables.go
+++ b/lang/blocktoattr/variables.go
@@ -33,7 +33,7 @@ func walkVariables(node dynblock.WalkVariablesNode, body hcl.Body, schema *confi
 	for _, child := range children {
 		if blockS, exists := schema.BlockTypes[child.BlockTypeName]; exists {
 			vars = append(vars, walkVariables(child.Node, child.Body(), &blockS.Block)...)
-		} else if attrS, exists := schema.Attributes[child.BlockTypeName]; exists && attrS.Type.ElementType().IsObjectType() {
+		} else if attrS, exists := schema.Attributes[child.BlockTypeName]; exists && attrS.Type.IsCollectionType() && attrS.Type.ElementType().IsObjectType() {
 			synthSchema := SchemaForCtyElementType(attrS.Type.ElementType())
 			vars = append(vars, walkVariables(child.Node, child.Body(), synthSchema)...)
 		}

